# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1464986825 14400
#      Fri Jun 03 16:47:05 2016 -0400
# Node ID 7b0436d0f5be5fbfb453009fb20e128253037499
# Parent  7f633f680f685f783bdcdc7654151de367ca2afe
rename maintainer-mode configure option to hd-id (bug #48101)

* configure.ac: Rename --enable-maintainer-mode configure option to
--enable-hg-id (enabled by default).

* Makefile.am (DIST_HOOKS): New variable.
(hg-id-dist-hook): Rename from maintainer-mode-dist-hook.
(dist-hook): Depend on $(DIST_HOOKS).
(all-local): Display HG ID for the build in summary message.
(HG-ID): Attempt to make warnings more visible.  Write
"hg-id-disabled" to HG-ID file if --disable-hg-id configure option was
used.  Warn if HG-ID is "hg-id-disabled" and build is not done from
Mercurial sources.

diff --git a/Makefile.am b/Makefile.am
--- a/Makefile.am
+++ b/Makefile.am
@@ -168,34 +168,40 @@ include src/module.mk
 include scripts/module.mk
 include doc/module.mk
 include doc/interpreter/images.mk
 include etc/module.mk
 include examples/module.mk
 include m4/module.mk
 include test/module.mk
 
-dist-hook: maintainer-mode-dist-hook doc-interpreter-dist-hook docs-dist-hook icons-dist-hook scripts-dist-hook
+DIST_HOOKS := \
+  doc-interpreter-dist-hook \
+  docs-dist-hook \
+  hg-id-dist-hook \
+  icons-dist-hook \
+  scripts-dist-hook
 
-if AMCOND_MAINTAINER_MODE
-## HG-ID must be correct to package distribution.
-maintainer-mode-dist-hook:
-	@echo "Maintainer-mode enabled.  Cannot package distribution!"
-	exit 1
-else
-maintainer-mode-dist-hook:
-endif
+dist-hook: $(DIST_HOOKS)
 
 if AMCOND_BUILD_DOCS
 docs-dist-hook:
 else
 docs-dist-hook:
 	@echo "Documentation disabled.  Cannot package distribution!" ; exit 1;
 endif
 
+if AMCOND_ENABLE_HG_ID
+hg-id-dist-hook:
+else
+hg-id-dist-hook:
+	@echo "WARNING: Octave was configured with --disable-hg-id" 1>&2
+endif
+.PHONY: hg-id-dist-hook
+
 if AMCOND_HAVE_ICON_TOOLS
 icons-dist-hook:
 else
 icons-dist-hook:
 	@echo "Packaging distribution requires icotool and rsvg-convert." ; exit 1;
 endif
 
 BUILT_SOURCES += \
@@ -238,16 +244,18 @@ ALL_LOCAL_TARGETS += \
 all-local: $(ALL_LOCAL_TARGETS)
 	@echo ""
 	@echo "Octave successfully built.  Now choose from the following:"
 	@echo ""
 	@echo "   ./run-octave    - to run in place to test before installing"
 	@echo "   make check      - to run the tests"
 	@echo "   make install    - to install (PREFIX=$(prefix))"
 	@echo ""
+	@echo "   HG ID for this build is \"`cat HG-ID`\""
+	@echo ""
 
 run-octave: run-octave.in build-aux/subst-script-vals.sh
 	$(AM_V_GEN)$(call simple-filter-rule,build-aux/subst-script-vals.sh) && \
 	chmod a+rx $@
 
 octave-config.h: config.h build-aux/mk-octave-config-h.sh
 	$(AM_V_GEN)$(SHELL) $(srcdir)/build-aux/mk-octave-config-h.sh $< > $@-t && \
 	$(simple_move_if_change_rule)
@@ -297,40 +305,43 @@ define changelog-from-hg-log
     mv $@-t $@; \
   fi
 endef
 
 ChangeLog:
 	$(AM_V_GEN)$(changelog-from-hg-log)
 .PHONY: ChangeLog
 
-if AMCOND_MAINTAINER_MODE
-## Avoid unnecessary building just because HG-ID changes (for example,
-## editing a .m file will cause + to be appended to the hg ID and that
-## will trigger compiling the build-info.cc files, relinking libraries
-## and .oct files).
-HG-ID:
-	$(AM_V_GEN)rm -f $@-t && \
-	echo "maintainer-mode" > $@-t && \
-	$(simple_move_if_change_rule)
-else
+if AMCOND_ENABLE_HG_ID
 HG-ID:
 ## Store hg ID in binaries so that we can identify botched installations
 ## that are using incorrect or mismatched libraries.
 	$(AM_V_GEN)rm -f $@-t && \
 	if [ -d $(srcdir)/.hg ]; then \
 	  ( cd $(srcdir) && hg identify --id || echo "unknown" ) > $@-t && \
 	  $(simple_move_if_change_rule); \
 	elif [ ! -f $(srcdir)/HG-ID ]; then \
-	  echo "$(srcdir)/HG-ID is missing!" 1>&2; \
+	  echo "WARNING: $(srcdir)/HG-ID is missing!" 1>&2; \
 	  echo "unknown" > $@-t && mv $@-t $@; \
 	else \
 	  echo "preserving existing HG-ID file" 1>&2; \
 	  $(cp_update_rule); \
+	  if [ "`cat $@`" = "hg-id-disabled" ]; then \
+	    echo "WARNING: HG-ID is 'hg-id-disabled'" 1>&2; \
+	  fi \
 	fi
+else
+## Avoid unnecessary building just because HG-ID changes (for example,
+## editing a .m file will cause "+" to be appended to the hg ID and that
+## will trigger compiling the build-info.cc files, relinking libraries
+## and .oct files).
+HG-ID:
+	$(AM_V_GEN)rm -f $@-t && \
+	echo "hg-id-disabled" > $@-t && \
+	$(simple_move_if_change_rule)
 endif
 .PHONY: HG-ID
 
 octetc_DATA += \
   CITATION \
   NEWS
 
 if AMCOND_INSTALL_BUILD_LOGS
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -280,23 +280,22 @@ case $with_shell in
   ;;
   *)
     SHELL_PATH=$with_shell
   ;;
 esac
 AC_DEFINE_UNQUOTED([SHELL_PATH], ["$SHELL_PATH"],
   [Define this to be the path to the shell command interpreter.])
 
-ENABLE_MAINTAINER_MODE=no
-AC_ARG_ENABLE([maintainer-mode],
-  [AS_HELP_STRING([--enable-maintainer-mode],
-    [enable maintainer mode features in build system])],
-  [if test "$enableval" = yes; then ENABLE_MAINTAINER_MODE=yes; fi], [])
-AM_CONDITIONAL([AMCOND_MAINTAINER_MODE],
-  [test $ENABLE_MAINTAINER_MODE = yes])
+ENABLE_HG_ID=yes
+AC_ARG_ENABLE([hg-id],
+  [AS_HELP_STRING([--disable-hg-id],
+    [disable embedding of hg id in libraries])],
+  [if test "$enableval" = no; then ENABLE_HG_ID=no; fi], [])
+AM_CONDITIONAL([AMCOND_ENABLE_HG_ID], [test $ENABLE_HG_ID = yes])
 
 ### Enable bounds checking on element references within Octave's array and
 ### matrix classes.  This slows down some operations a bit, so it is turned off
 ### by default.
 
 ENABLE_BOUNDS_CHECK=no
 AC_ARG_ENABLE([bounds-check],
   [AS_HELP_STRING([--enable-bounds-check],
